### PR TITLE
Eat closest dot behavior

### LINKED ISSUE #25: Add eatClosestPill behavior

### PR DESCRIPTION
**Pull request by ll-nick**
Closes #25 

This adds the `EatClosestDotBehavior`.

The behavior itself is pretty simple as the heavy lifting is done within the A* class. Here's a brief summary of the changes:
- There is now a getter for the TileType in the maze class. Before, it was only possible to check for walls, not for dots.
- Add a predecessor to the cell class allowing to generate a path by traversing from the goal to the start via the predecessor property
- Add a shortestPath function. computeDistance is now simply a wrapper that returns the length of the returned path
- Add another interface to the A* class computing the path to the closest dot reusing most of the existing A* logic
- Continuously update the maze withing the A* class since we now need the dynamic info about dot positions. It's just a pointer so it shouldn't be a big issue.
- Add the closestDotBehavior to the Pacman Agent. For now I just added it to the main priority arbitrator. We need to change that after adding #27 
- Yet another rename for clarity: the A* distance is now called mazeDistance because I replaced the function that used to return the euclidean distance by the manhattan distance since it's a better heuristic for A*.

I think that's about it. A couple of points I'm unsure about:
- Should we also consider power pellets here or do we do that separately or not at all?
- Is there a sensible heuristic for the closest dot path? Right now I just use 0 which is obviously a crappy heuristic but I couldn't think of anything else and I'm not sure if there is a better solution.

### CODE REVIEW

File: demo/src/astar.cpp [hunk 1]
```diff
@@ -17,44 +17,83 @@ int nonNegativeModulus(const int& numerator, const int& denominator) {
...
+    }
+    startCell.distanceFromStart = 0;
+    HeuristicFunction heuristic = [](const Cell&) { return 0; };
+    startCell.heuristic = 0;
+
+    openSet.push(startCell);
+
+    while (!openSet.empty()) {
+        // Unfortunately, the pacman simulation will handle the dot consumption after the move, therefore we need to
+        // explicitely exclude the start position from the search.
```
Comment (orzechow) [inline demo/src/astar.cpp:86]:
```suggestion
        // explicitly exclude the start position from the search.
```

File: demo/test/astar.cpp [hunk 2]
```diff
@@ -63,27 +63,83 @@ TEST_F(AStarTest, distanceWithVerticalTunnel) {
...
+    environmentModel_->setMaze({5, 5}, str);
+
+    AStar astar(environmentModel_->maze());
+    Path path = astar.shortestPath({2, 1}, {3, 3});
+    Path targetPath = {Direction::LEFT, Direction::DOWN, Direction::DOWN, Direction::RIGHT, Direction::RIGHT};
+    ASSERT_EQ(path.size(), targetPath.size());
+    for (int i = 0; i < targetPath.size(); i++) {
+        EXPECT_EQ(path.at(i), targetPath.at(i));
+    }
+}
```
Comment (orzechow) [inline demo/test/astar.cpp:104]: :heart_eyes:

**Review by orzechow — APPROVED** [review]
Thanks! Looks good!

Some minor suggestions, feel free to consider them and merge right away

File: demo/include/demo/eat_closest_dot_behavior.hpp [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        }
+
+        return Command{pathToClosestDot};
+    }
+
+    bool checkInvocationCondition(const Time& /*time*/) const override {
+        // This behavior is only applicable if there is at least one dot. We should check for the presence of a dot
+        // here, but since the game is won when all dots are collected, we assume at least one dot exists when this
+        // behavior is invoked.
+        return true;
```
[the remark this answers is not in the document]
Comment (ll-nick) [reply demo/include/demo/eat_closest_dot_behavior.hpp:38]: True I guess. I'll add a check :+1:

**Pull request merged**